File: external-declarations/src/compiler/debug.ts
type AnyFunction = (...a: any) => any;
/** @internal */
export namespace Debug {
    export function fail(message?: string, stackCrawlMark?: AnyFunction): never {
        debugger;
        const e = new Error(message ? `Debug Failure. ${message}` : "Debug Failure.");
        if ((Error as any).captureStackTrace) {
            (Error as any).captureStackTrace(e, stackCrawlMark || fail);
        }
        throw e;
    }

    export function assert(expression: unknown, message?: string, verboseDebugInfo?: string | (() => string), stackCrawlMark?: AnyFunction): asserts expression {
        if (!expression) {
            message = message ? `False expression: ${message}` : "False expression.";
            if (verboseDebugInfo) {
                message += "\r\nVerbose Debug Information: " + (typeof verboseDebugInfo === "string" ? verboseDebugInfo : verboseDebugInfo());
            }
            fail(message, stackCrawlMark || assert);
        }
    }

    export function assertNever(member: never, message = "Illegal value:", stackCrawlMark?: AnyFunction): never {
        const detail = JSON.stringify(member);
        return fail(`${message} ${detail}`, stackCrawlMark || assertNever);
    }
}

File: tsc/internal/debug/debug.go
package debug

import (
	"fmt"
)

func Fail(reason string) {
	if len(reason) == 0 {
		reason = "Debug failure."
	} else {
		reason = "Debug failure. " + reason
	}
	// runtime.Breakpoint()
	panic(reason)
}

func FailBadSyntaxKind(node interface{ KindString() string }, message ...any) {
	var msg string
	if len(message) == 0 {
		msg = "Unexpected node."
	} else {
		msg = fmt.Sprint(message...)
	}
	Fail(fmt.Sprintf("%s\nNode %s was unexpected.", msg, node.KindString()))
}

func AssertNever(member any, message ...any) {
	var msg string
	if len(message) == 0 {
		msg = "Illegal value:"
	} else {
		msg = fmt.Sprint(message...)
	}
	var detail string
	if m, ok := member.(interface{ KindString() string }); ok {
		detail = m.KindString()
	} else if m, ok := member.(fmt.Stringer); ok {
		detail = m.String()
	} else {
		detail = fmt.Sprintf("%v", member)
	}
	Fail(fmt.Sprintf("%s %s", msg, detail))
}

func Assert(value bool, message ...any) {
	if value {
		return
	}
	assertSlow(message...)
}

func assertSlow(message ...any) {
	// See https://dave.cheney.net/2020/05/02/mid-stack-inlining-in-go
	var msg string
	if len(message) > 0 {
		msg = "False expression: " + fmt.Sprint(message...)
	} else {
		msg = "False expression."
	}
	Fail(msg)
}
